i6_models/config.py

```diff
@@ -51,3 +53,18 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+@dataclass
+class SubassemblyWithOptions:
+    """
+    Dataclass for a combination of an Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Callable[[ModelConfiguration], nn.Module]
+    cfg: ModelConfiguration
+
+    def construct(self) -> nn.Module:
+        """Constructs an instance of the given module class"""
+        return self.module_class(self.cfg)
```

Comment:
```suggestion
ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
ModuleType = TypeVar("ModuleType", bound=nn.Module)


@dataclass
class SubassemblyWithOptions(Generic[ConfigType, ModuleType]):
    """
    Dataclass for a combination of an Subassembly/Part and the corresponding configuration.
    Also provides a function to construct the corresponding object through this dataclass
    """

    module_class: Callable[[ConfigType], ModuleType]
    cfg: ConfigType

    def construct(self) -> ModuleType:
        """Constructs an instance of the given module class"""
        return self.module_class(self.cfg)
```

The original version raises typing issues because concrete instances of module classes only accept subclasses of `ModelConfiguration` and not `ModelConfiguration` itself. This should fix typing issues.